syntax = "proto2";

package iit.advr;


message Time
{
  required int32 sec = 1;
  required int32 nsec = 2;
}

message Header
{
  optional string str_id = 1;
  optional Time stamp    = 2;
  optional int32 index   = 3;
}

message Vector2d
{
  required double x = 1;
  required double y = 2;
}

///////////////////////////////////////////////////////////////////////////////

message Trajectory_cmd {

    enum Type { 
        SINE        = 1;
        COSINE      = 2;
        STEP        = 3;
        SMOOTHER    = 4;
        SPLINE      = 5;
        HOME        = 6;
    }
    
    required Type       type = 1;
    required string     name = 2;
    required int32      board_id = 3;
    repeated double     x = 4;
    repeated double     y = 5;
}

message Gains {

    enum Type { 
        POSITION    = 1;
        IMPEDANCE   = 2;
        TORQUE      = 3;
    }
    // Identifies which field is filled in.
    required Type   type = 1;
    
    required double kp  = 2;
    required double ki  = 3;
    required double kd  = 4;
    optional double kkp = 5;
    optional double kki = 6;
    optional double kkd = 7;
    
}

message Repl_hdr {

}

message Ctrl_cmd {
    
    enum Type {
        CTRL_TEST_DONE          = 1;
        CTRL_TEST_ERROR         = 2;
        CTRL_DAC_TUNE           = 3;
        CTRL_REMOVE_TORQUE_OFFS = 4;
        CTRL_SET_ZERO_POSITION  = 5;
        CTRL_FAN                = 6;
        CTRL_LED                = 7;
        CTRL_SANDBOX            = 8;
        CTRL_REF_FILTER         = 9;
        CTRL_POWER_MOD          = 10;
        CTRL_CMD_START          = 11;
        CTRL_CMD_STOP           = 12;
        
    }
    // Identifies which field is filled in.
    required Type   type        = 1;
    required int32  board_id    = 2;

    optional float  value       = 3;
}

message Flash_cmd {

    enum Type { 
        SAVE_PARAMS_TO_FLASH    = 1;
        LOAD_PARAMS_FROM_FLASH  = 2;
        LOAD_DEFAULT_PARAMS     = 3;
        
    }
    // Identifies which field is filled in.
    required Type   type        = 1;
    required int32  board_id    = 2;
}

message Ecat_Master_cmd {

    enum Type { 
        START_MASTER        = 1;
        STOP_MASTER         = 2;
        GET_SLAVES_DESCR    = 3;
        
    }
    // Identifies which field is filled in.
    required Type   type    = 1;
}

message FOE_Master {

    required string filename    = 1;
    required uint32 password    = 2; 
    optional string mcu_type    = 3;
    optional int32  slave_pos   = 4;
    optional int32  board_id    = 5;
}

///////////////////////////////////////////////////////////////////////////////
//
///////////////////////////////////////////////////////////////////////////////

enum CmdType { 
    SEND_TRAJECTORY     = 1;
    SET_GAINS           = 2;
    CTRL_CMD            = 3;
    FLASH_CMD           = 4;
    ECAT_MASTER_CMD     = 5;    
    FOE_MASTER          = 6;
    HOMING              = 7;
}

message Repl_cmd {

    // Identifies which field is filled in.
    required CmdType            type = 1;
    
    optional Header             header = 2;
    
    // One of the following will be filled in.
    optional Trajectory_cmd     trajectory_cmd  = 3;
    optional Gains              gains           = 4;
    optional Ctrl_cmd           ctrl_cmd        = 5;
    optional Flash_cmd          flash_cmd       = 6;
    optional Ecat_Master_cmd    ecat_master_cmd = 7;
    optional FOE_Master         foe_master      = 8;
    
}

message Cmd_reply {

   enum Type { 
        ACK  = 1;
        NACK = 2;
    }
    // Identifies which field is filled in.
    required Type           type        = 1;
    required CmdType        cmd_type    = 2;
    required string         msg         = 3;
    
    optional Header         header      = 4;
    
}
